src/stream/src/executor/top_n_global.rs

```diff
@@ -0,0 +1,200 @@
+    ) -> StreamExecutorResult<Self> {
+        let info = input.info();
+        let schema = input.schema().clone();
+
+        Ok(TopNExecutorWrapper {
+            input,
+            inner: InnerTopNGlobalExecutor::new(
+                info,
+                schema,
+                order_pairs,
+                offset_and_limit,
+                pk_indices,
+                total_count,
+                executor_id,
+            )?,
+        })
+    }
+}
+
+pub struct InnerTopNGlobalExecutor{
+    info: ExecutorInfo,
+
+    /// Schema of the executor.
+    schema: Schema,
+
+    /// `LIMIT XXX`. None means no limit.
+    limit: Option<usize>,
+
+    /// `OFFSET XXX`. `0` means no offset.
+    offset: usize,
+
+    /// The primary key indices of the `TopNExecutor`
+    pk_indices: PkIndices,
+
+    /// The internal key indices of the `TopNExecutor`
+    internal_key_indices: PkIndices,
+
+    /// The order of internal keys of the `TopNExecutor`
+    internal_key_order_types: Vec<OrderType>,
+
+    /// The output row num of the `TopNExecutor`
+    total_count: usize,
+}
+
+pub fn generate_internal_key(
+    order_pairs: &[OrderPair],
+    pk_indices: PkIndicesRef,
+) -> (PkIndices,  Vec<OrderType>) {
+    let mut internal_key_indices = vec![];
+    let mut internal_order_types = vec![];
+
+    for order_pair in order_pairs {
+        internal_key_indices.push(order_pair.column_idx);
+        internal_order_types.push(order_pair.order_type);
+    }
+    for pk_index in pk_indices {
+        if !internal_key_indices.contains(pk_index) {
+            internal_key_indices.push(*pk_index);
+            internal_order_types.push(OrderType::Ascending);
+        }
+    }
+
+    (
+        internal_key_indices,
+        internal_order_types,
+    )
+}
+
+impl InnerTopNGlobalExecutor {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        input_info: ExecutorInfo,
+        schema: Schema,
+        order_pairs: Vec<OrderPair>,
+        offset_and_limit: (usize, Option<usize>),
+        pk_indices: PkIndices,
+        total_count: usize,
+        executor_id: u64,
+    ) -> StreamExecutorResult<Self> {
+        let (internal_key_indices, internal_key_order_types) =
+            generate_internal_key(&order_pairs, &pk_indices);
+
+        Ok(Self{
+            info: ExecutorInfo {
+                schema: input_info.schema,
+                pk_indices: input_info.pk_indices,
+                identity: format!("TopNGlobalExecutor {:X}", executor_id),
+            },
+            schema,
+            offset: offset_and_limit.0,
+            limit: offset_and_limit.1,
+            pk_indices,
+            internal_key_indices,
+            internal_key_order_types,
+            total_count,
+        })
+    }
+    /// Sort `new_row` and keep only `self.limits` entries
+    fn sort_and_filter(&self, new_rows: &Vec<Row>, schema: &Schema) {
+        let limit = self.limit.unwrap_or(0);
+        let offset = self.offset;
+
```

Comment:
You may refer to https://github.com/singularity-data/risingwave/blob/761ec04f9eb1296c326ffff0942cf4d9787aca6d/src/stream/src/executor/managed_state/top_n/top_n_state_new.rs#L115-L124 if you want to implement something similar in the future. We can get the storage pk from the deserialized row by re-using `get_topn_row`.

Reply:
it seems that `row` have been sorted before?

Reply:
and the `ordered` just flips  the order if needed？

Reply:
You'll need to merge multiple batches of sorted `Row`s by users' specified order (like `a ASC, b DESC`). Therefore, to merge them, we'll need to use `OrderedRow` to flip the order of some columns, so as to produce correct result.